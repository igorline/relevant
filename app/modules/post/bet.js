import React, { useState, Fragment, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import {
  View,
  Text,
  HoverButton,
  SecondaryText,
  LinkFont,
  SmallText,
  Header,
  Divider
} from 'modules/styled/uni';
import { getTimestamp, abbreviateNumber as toFixed } from 'utils/numbers';
import { colors } from 'styles';
import { VOTE_COST_RATIO } from 'server/config/globalConstants';
import CoinStat from 'modules/stats/coinStat.component';
import { bet, getPostInvestments } from 'modules/post/invest.actions';
import { hideModal } from 'modules/navigation/navigation.actions';
import { computeShares } from 'app/utils/post';
import { computePostPayout } from 'app/utils/rewards';
import { useCommunity } from 'modules/community/community.selectors';
import { PieChart } from 'modules/stats/piechart';
import Tooltip from 'modules/tooltip/tooltip.component';
import { exchangeLink } from 'modules/wallet/price.context';
import ULink from 'modules/navigation/ULink.component';
import { NotificationToggle } from 'modules/profile/settings/settings.toggle';

export function BetContainer() {
  const user = useSelector(state => state.auth.user) || {};
  const post = useSelector(state => state.posts.posts[state.navigation.modalData.postId]);
  if (!user || !post) return null;
  return <Bet user={user} post={post} />;
}

Bet.propTypes = {
  user: PropTypes.object,
  post: PropTypes.object
};

function Bet({ user, post }) {
  const dispatch = useDispatch();
  const earning = useSelector(state =>
    state.earnings.pending
      .map(e => state.earnings.entities[e])
      .find(ee => ee.post === post._id)
  );

  const title = earning ? 'Increase Your Bet' : 'Bet on the Relevance of this Post';

  const totalBalance = user.balance + user.tokenBalance;
  const maxBet = totalBalance - user.lockedTokens;
  const defaultAmount = Math.max(maxBet * VOTE_COST_RATIO, 0);
  const [amount, setAmount] = useState(defaultAmount);

  if (!user || !post) return null;

  const time = getTimestamp(post.data.payoutTime).toLowerCase();

  const plusAmount = () =>
    setAmount(a => {
      const largeStep = (maxBet - defaultAmount) / 5;
      const smallStep = defaultAmount / 5;
      const err = defaultAmount / 100;
      return a + smallStep <= defaultAmount + err
        ? a + smallStep
        : Math.min(a + largeStep, maxBet);
    });

  const minusAmount = () =>
    setAmount(a => {
      const largeStep = (totalBalance - user.lockedTokens - defaultAmount) / 5;
      const smallStep = defaultAmount / 5;
      const err = defaultAmount / 100;
      if (a - smallStep <= 0 + err) return a;
      return a - largeStep >= defaultAmount - err
        ? a - largeStep
        : Math.max(0, a - smallStep);
    });

  const placeBet = () => {
    dispatch(hideModal());
    dispatch(bet({ postId: post._id, stakedTokens: amount }));
  };

  const power = (100 * amount) / maxBet;

  const tooltipData = {
    text:
      'Posts that get upvoted by lots of users with high Reputation get payouts.\n\nBet more coins and bet early in order to win the biggest portion of the payout.'
  };

  return (
    <View>
      <Header inline={1} mr={2}>
        {title} <Tooltip inline={1} name={'betInfo'} data={tooltipData} info />
      </Header>

      <View fdirection={'row'} align={'baseline'}>
        <SmallText mt={1} mr={1}>
          Payout: {time}
        </SmallText>
      </View>

      <View mt={4} fdirection="row" justify="space-between" align={'center'}>
        <CircleButton onPress={minusAmount}>â€“</CircleButton>
        <View pr={1}>
          <CoinStat
            align={'center'}
            fs={4.5}
            lh={4.5}
            spaceBetween={1}
            c={colors.black}
            amount={amount}
          />
        </View>
        <CircleButton onPress={plusAmount}>+</CircleButton>
      </View>

      <View mt={2}>
        <View h={0.5} fdirection={'row'}>
          <View w={`${power}%`} bg={colors.blue} />
          <View w={`${100 - power}%`} bg={colors.lightBorder} />
        </View>
        <SecondaryText alignself={'center'} mt={0.5}>
          Available Coins: <SmallCoinStat amount={maxBet - amount} />
        </SecondaryText>
        <SecondaryText alignself={'center'} mt={0.5}>
          <ULink to={exchangeLink()} external target="_blank">
            Get more coins
          </ULink>
        </SecondaryText>
      </View>

      <View mt={4}>
        <PotentialRewards maxBet={maxBet} post={post} amount={amount} earning={earning} />
      </View>

      <HoverButton mt={3} onPress={placeBet}>
        Bet {toFixed(amount)} Coins
      </HoverButton>

      <SmallText mt={2}>
        *You get your coins back once the betting round ends, win or lose.
      </SmallText>

      <Divider mt={2} />

      <NotificationToggle
        notification={user.notificationSettings.bet.manual}
        parent={'bet'}
        label={'manual'}
        togglePosition={'right'}
        DescriptionComponent={SmallText}
        text={{
          description: "Don't like betting?\nYou can disable the manual betting mode."
        }}
      />

      {/*      <SmallText mt={2}>
        If this post ranks highly you will win some coins, either way you allways get your
        coins back.
      </SmallText> */}
    </View>
  );
}

PotentialRewards.propTypes = {
  post: PropTypes.object,
  earning: PropTypes.object,
  amount: PropTypes.number,
  maxBet: PropTypes.number
};

function PotentialRewards({ post, amount, earning, maxBet }) {
  const community = useCommunity();

  const investments = useSelector(state =>
    (state.investments.posts[post._id] || [])
      .map(_id => state.investments.investments[_id])
      .filter(inv => inv.stakedTokens > 0 && inv.investor !== state.auth.user._id)
  );

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getPostInvestments(post._id));
  }, [dispatch, post._id]);

  const existingShares = earning ? earning.shares : 0;
  const existingStake = earning ? earning.stakedTokens : 0;

  const bets = investments.length;
  const users = bets > 1 ? 'users' : 'user';
  const invText = bets
    ? `${bets} ${users} bet a total of ${toFixed(
        post.data.totalShares - existingStake
      )} coins on this post`
    : 'You are the first to bet on this post!';

  const shares = computeShares({ post, stakedTokens: amount });
  const postRewards = computePostPayout(post.data, community);
  const shareOfRewards = !bets
    ? amount / maxBet
    : (shares + existingShares) / (post.data.shares + shares);
  const shareOfRewardsPercent = shareOfRewards * 100;
  const potentialRewards = postRewards * shareOfRewards;
  // const showPie = shareOfRewards !== 1;

  const shareEl = (
    <SmallText inline={1}>
      {'  '}
      <Text style={{ top: 2 }} inline={1} mb={-0.5}>
        <PieChart
          w={'12px'}
          h={'12px'}
          percent={100 - shareOfRewardsPercent}
          strokeWidth={30}
          color={colors.blue}
        />
      </Text>{' '}
      {toFixed(shareOfRewardsPercent)}%
    </SmallText>
  );

  return (
    <Fragment>
      <SmallText>{invText}</SmallText>
      {potentialRewards > 0 && (
        <SmallText inline={1} mt={0.25}>
          Your estimated rewards: <SmallCoinStat showPrice amount={potentialRewards} />
          {shareEl}
        </SmallText>
      )}
    </Fragment>
  );
}

CircleButton.propTypes = {
  children: PropTypes.node,
  onPress: PropTypes.func
};

function CircleButton({ children, onPress }) {
  return (
    <HoverButton
      border={1}
      bc={colors.blue}
      w={8}
      h={8}
      minwidth={'auto'}
      bradius={4}
      bg={colors.white}
      onPress={onPress}
    >
      <LinkFont c={colors.blue} fs={2.5} lh={2.5}>
        {children}
      </LinkFont>
    </HoverButton>
  );
}

SmallCoinStat.propTypes = {
  amount: PropTypes.number
};

function SmallCoinStat({ amount, ...rest }) {
  return (
    <CoinStat
      size={1.5}
      fs={1.5}
      secondary
      c={colors.black}
      inline={1}
      amount={amount}
      {...rest}
    />
  );
}

export default BetContainer;
