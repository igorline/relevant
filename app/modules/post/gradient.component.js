import React from 'react';
import PropTypes from 'prop-types';
import LinearGradient from 'react-native-linear-gradient';

const bgGradient = [
  'hsla(240, 70%, 30%, .03)',
  'hsla(240, 70%, 20%, .08)',
  'hsla(240, 70%, 10%, .4)',
  'hsla(240, 70%, 8%, .7)',
  'hsla(240, 70%, 5%, .6)'
];

const previewGradient = [
  'hsla(240, 70%, 10%, .4)',
  'hsla(240, 70%, 8%, .7)',
  'hsla(240, 70%, 5%, .6)'
];

export default function Gradient({ title, image, preview }) {
  const textOverlay = preview ? previewGradient : bgGradient;
  const colors = image ? textOverlay : generateColors(title);

  const start = image ? { x: 0.5, y: 0.0 } : { x: 0.8, y: 0.0 };
  const end = image ? { x: 0.5, y: 1.0 } : { x: 0.2, y: 1.0 };

  return <LinearGradient start={start} end={end} colors={colors} style={{ flex: 1 }} />;
}

function generateColors(string) {
  let color = string ? string.length : 0;
  color = (color % 220) + 200 || 200;
  color = Math.max(100, color);
  return [
    `hsla(${parseInt(color - 30, 10)}, 100%, 50%, 1)`,
    `hsla(${parseInt(color, 10)},      100%, 50%, 1)`,
    `hsla(${parseInt(color + 30, 10)}, 100%, 50%, 1)`
  ];
}

Gradient.propTypes = {
  preview: PropTypes.bool,
  image: PropTypes.bool,
  title: PropTypes.string
};
