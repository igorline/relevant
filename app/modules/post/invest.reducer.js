import * as types from 'core/actionTypes';

const initialState = {
  userInvestments: {},
  investments: {},
  myInvestments: [],
  myEarnings: {},
  posts: {},
  loaded: {},
  loadedProfileInv: false,
  voteSuccess: null
};

export default function investments(state = initialState, action) {
  switch (action.type) {
    case types.UPDATE_POST_INVESTMENTS: {
      return {
        ...state,
        voteSuccess: action.payload
      };
    }

    case types.LOADING_POST_INVESTMENTS: {
      return {
        ...state,
        loaded: {
          ...state.loaded,
          [action.payload]: false
        }
      };
    }

    case types.SET_POST_INVESTMENTS: {
      const post = state.posts[action.payload.postId] || [];
      return {
        ...state,
        posts: {
          ...state.posts,
          [action.payload.postId]: [
            ...post.slice(0, action.payload.index),
            ...action.payload.data.result.investments
          ]
        },
        loaded: {
          ...state.loaded,
          [action.payload.postId]: true
        },
        investments: {
          ...state.investments,
          ...action.payload.data.entities.investments
        }
      };
    }

    case types.UPDATE_EARNINGS: {
      const earnings = {};
      action.payload.forEach(earning => {
        earnings[earning.post] = earning;
      });
      return {
        ...state,
        myEarnings: {
          ...state.myEarnings,
          ...earnings
        }
      };
    }

    case 'SET_INVESTMENTS': {
      const id = action.payload.userId;
      const currentInvestments = state.userInvestments[id] || [];
      return {
        ...state,
        userInvestments: {
          ...state.userInvestments,
          [id]: [
            ...currentInvestments.slice(0, action.payload.index),
            ...action.payload.investments.result.investments
          ]
        },
        investments: {
          ...state.investments,
          ...action.payload.investments.entities.investments
        },
        loadedProfileInv: true
      };
    }

    case 'LOADING_INVESTMENTS': {
      return {
        ...state,
        loadedProfileInv: false
      };
    }

    case types.LOGOUT_USER: {
      return { ...initialState };
    }

    default:
      return state;
  }
}
