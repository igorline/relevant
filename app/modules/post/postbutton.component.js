import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Image, Touchable } from 'modules/styled/uni';
import styled from 'styled-components/primitives';
import { isNative } from 'styles';

const BUTTON_IMAGES = {
  UPVOTE: {
    white: require('app/public/img/vote_buttons/upvote-white.png'),
    default: require('app/public/img/vote_buttons/upvote-grey-outline.png'),
    active: require('app/public/img/vote_buttons/upvote-blue.png'),
    hover: require('app/public/img/vote_buttons/upvote-blue-outline.png')
  },
  DOWNVOTE: {
    white: require('app/public/img/vote_buttons/downvote-white.png'),
    default: require('app/public/img/vote_buttons/downvote-grey-outline.png'),
    active: require('app/public/img/vote_buttons/downvote-red.png'),
    hover: require('app/public/img/vote_buttons/downvote-red-outline.png')
  }
};

const ButtonImage = styled(Image)`
  ${() =>
    !isNative
      ? `transition-property: all;
        transition-duration: 0.2s;
        transition-timing-function: ease`
      : ''}
  ${p =>
    p.hover && !p.active && !p.disabled && !isNative ? 'transform: scale(1.1);' : ''}
`;

PostButton.propTypes = {
  color: PropTypes.string,
  alt: PropTypes.string,
  isActive: PropTypes.bool,
  imageSet: PropTypes.oneOf(['DOWNVOTE', 'UPVOTE']),
  onPress: PropTypes.func
};

function PostButton({ alt, isActive, imageSet, onPress, color }) {
  const [hover, setHover] = useState(false);
  const [active, setActive] = useState(false);

  const images = BUTTON_IMAGES[imageSet];
  const defaultState = images[color] || images.default;
  const source = isActive ? images.active : hover ? images.hover : defaultState;

  return (
    <Touchable onPress={e => onPress(e)} bradius={2}>
      <ButtonImage
        w={3}
        h={3}
        alt={alt}
        resizeMode="contain"
        source={source}
        hover={hover}
        active={active}
        onMouseDown={() => setActive(1)}
        onMouseUp={() => setActive(0)}
        onMouseEnter={() => setHover(1)}
        onMouseLeave={() => {
          setHover(0);
          setActive(0);
        }}
      />
    </Touchable>
  );
}

export default PostButton;
