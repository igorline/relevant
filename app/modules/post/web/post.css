@import 'styles/global.css';

.postContainer {
  /*max-width: 800px;*/
  flex: 1;
  display: flex;
  min-width: 0;
  flex-direction: column;
  /*align-self: center;*/
  padding-bottom: 100px;
  &.narrow {
    max-width: 600px;
    margin: auto;
  }
}

.shadowBox {
  /*border: 1px solid #000;*/
  /*box-shadow: 2px 2px 0 0 #000;*/
}

.close {
  position: absolute;
  top: 5px;
  right: 10px;

  background-image: url('/img/x.png');
  background-size: cover;
  pointer-events: none;
  div {
    pointer-events: auto;
    width: 15px;
    height: 15px;
    cursor: pointer;
  }
}

/*.singlePost .postContainer  {
  margin: 0 0 0 $leftMargin;
}*/

.postinfo {
  margin-bottom: 10px;
  position: relative;
  &.shadowBox {
    display: flex;
    flex-direction: row;
    width: 100%;
    padding: 20px;
    min-height: 300px;
    background-position: center;
    color: white;
    background-size: cover;
  }
  &.minimalPost{
    flex-direction: row;
    color: black;
    display: flex;
  }

  .image {
    position: relative;
    /*height: 150px;*/
    /*align-self: flex-start;*/
    div {
      background-size: cover;
      background-position: center;
      /*width: 150px;*/
      min-height: 120px;
      height: 100%;
      /*float: right;*/
      width: 100%;
    }
  }


  .headlineContainer {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    flex: 1;
    margin-left: 10px;
    border-bottom: 3px solid;
  }
  .headline {
    font-size: 28px;
    line-height: 30px;
    margin: 0 0;
    display: inline-block;
    padding-top: 5px;
    padding-right: 20px;
    /*color: $blue;*/
  }

  &:hover {
    .headlineContainer {
      border-color: $blue;
    }
    .domain {
      color: blue;
    }
    .headline {
      color: $blue;
    }
  };

  .headline:after {
    content: '';
    display: inline-block;
    width: 10px;
    height: 10px;
    margin-left: -10px;
    background-image: url(/img/linkout.png);
    background-size: cover;
    transform: translateX(150%);
  }
  .domain {
    font-size: 15px;
    padding-bottom: 5px;
  }
  &.small {
    max-height: none;
    .image {
      width: 200px;
      height: 100px;
      > div {
        height: 100px;
        min-height: 100px;
        position: relative;
      }
    }
    .headline {
      font-size: 25px;
      line-height: normal;
    }
    .headlineContainer {
      flex: 1;
      /*border: none;*/
    }
  }
  &.preview {
    padding: 0px;
    border: none;
    box-shadow: 0px 0px 0px 1px $borderGrey;
    align-items: center;
    &:hover {
      box-shadow: 0px 0px 0px 1px $borderGrey;
    };
    .image {
      width: 90px;
      height: auto;
      margin-right: 0px;
      div {
        min-height: 60px;
        height: 60px;
        position: relative;
        /*display: absolute;*/
      }
    }
    .headlineContainer {
      border: none;
      padding-left: 0;
    }
    .headline {
      font-family: Arial;
      font-weight: 100;
      color: $greyText;
      font-size: 13px;
      padding: 5px 10px 0 10px;
      line-height: normal;
      &::after {
        display: none;
      }
    }
    .domain {
      font-size: 10px;
      padding-bottom: 0px;
      color: $greyText;
      padding: 0 10px;
    }
  }
}

.post {
  .image, .postMeta {
    width: 200px;
    height: auto;

    /*align-self: flex-start;*/
    justify-content: flex-end;
  }
  .postMeta {
    flex: 1;
    margin-right: 20px;
  }
}
.post {
  .postBody, .headlineContainer {
    flex: 3;
  }
}

.post {
  .postMeta {
    display: flex;
    flex-direction: row;
    align-items: flex-start;
    justify-content: flex-end;
    width: 0px;
    .avatarBox.repost {
      margin-left: 10px;
      .avatar {
        width: 26px;
        height: 26px;
        flex: 0 0 26px;
      }
    }
  }
}

.post {
  background-color: white;
  margin: 0px 0 20px 0;
  position: relative;
  padding: 25px 20px;
  box-shadow: $postShadow;
  cursor: pointer;
  &:hover {
    box-shadow: $postShadowHover;
  };
  /*@extend %flatBorderBottom;*/
  /*border-bottom: 1px solid $borderGrey;*/
}

.postContent {
  padding-top: 10px;
  display: flex;
  padding-left: 10px;
}

.post .avatarBox {
  margin: 10px 0 20px 0;
}

.postBody {
  flex: 1;
  /*margin-top: -10px;*/
  /*border-left: 6px solid $borderGrey;*/
  /*padding-left: 50px;*/
  font-family: 'Georgia', sans-serif;
  font-size: 17px;
  line-height: 1.52;
  color: @darkGrey;
  flex: 1;
  padding-bottom: 10px;
  /*margin-left: 50px;*/

  /*border-bottom: 1px solid $borderGrey;*/
}
.postBody span:after,
.postBody a:after {
  content: '\0020';
}
.repostBody {
  /*font-size: 14px;*/
  line-height: 21px;
  padding-left: 10px;
  border-left: 1px solid #b4b4b5;
  margin: 20px 0 10px 0;
}

.repostText {
  font-size: 16px;
}

.postbuttons {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin-top: 25px;
  margin-bottom: 5px;
  font-family: Arial, sans-serif;
  .right .details {
    display: flex;
    align-items: center;
  }
  .left {
    display: flex;
    align-items: center;
    justify-content: center;
  }
  .left > a {
    margin-right: 20px;
  }
  .upvote, .downvote {
    width: 25px;
    height: 25px;
  }
  .fraction {
    display: inline-block;
    margin-right: 20px;
  }
  a {
    display: flex;
    /*box-shadow: 0 0 10px -4px $grey;*/
  }
}

.fraction {
  font-size: 12px;
  color: #9C9C9C;
  text-align: center;
  min-width: 30px;
  img {
    margin-left: 1px;
    margin-right: 2px;
    margin-bottom: -1px;
    height: 11px;
    width: auto;
  }
  .num {
    border-bottom: 1px solid;
  }
  .dem {
    padding-bottom: 4px;
  }
}

.commentcount {
  display: inline-block;
  font-size: 10px;
  color: #9c9c9c;
  margin-right: 20px;
  cursor: pointer;
  img {
    margin-right: 4px;
  }
  span {
    position: relative;
  }
}

.post .share {
  flex: 0 0 21px;
  width: 21px;
  height: 18px;
}

@media screen
  and (max-width: 736px) {

  .postContainer {
    padding: 0;
  }
  .post .shadowBox {
    /*padding: 0;*/
    /*position: relative;*/
    /*border: 0;*/
    /*min-height: 240px;*/
    /*box-shadow: 0 0;*/
    /*margin-bottom: 0;*/
  }
  .post .image {
/*    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 0;
    background-color: #212121;*/
  }
  .post .shadowBox > div {
    width: 100%;
    padding: 10px;
    z-index: 1;
    color: white;
    flex-direction: column;
    display: flex;
    justify-content: flex-end;
    background: linear-gradient(0deg, #242425, rgba(36,36,36,0.3));
  }
  .post .headline {
    font-size: 34px;
  }
  .postContent {
    padding: 0px 10px;
  }
}
