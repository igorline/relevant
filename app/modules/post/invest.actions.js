import { normalize, schema } from 'normalizr';
import {
  UPDATE_POST_INVESTMENTS,
  UNDO_POST_INVESTMENT,
  SET_POST_INVESTMENTS,
  SET_INVESTMENTS,
  LOADING_INVESTMENTS,
  SET_USERS,
  LOADING_POST_INVESTMENTS
} from 'core/actionTypes';
import { api, alert } from 'app/utils';
import { setPostsSimple } from 'modules/post/post.actions';
import {
  showPushNotificationPrompt,
  showBetPrompt
} from 'modules/activity/activity.actions';

const Alert = alert.Alert();

const linkSchema = new schema.Entity('links', {}, { idAttribute: '_id' });

const postSchema = new schema.Entity(
  'posts',
  { metaPost: linkSchema },
  { idAttribute: '_id' }
);

const userSchema = new schema.Entity('users', {}, { idAttribute: '_id' });

const investmentSchema = new schema.Entity(
  'investments',
  {
    post: postSchema,
    investor: userSchema
  },
  { idAttribute: '_id' }
);

export function updatePostVote(voteObj) {
  return {
    type: UPDATE_POST_INVESTMENTS,
    payload: voteObj
  };
}

export function undoPostVote(post) {
  return {
    type: UNDO_POST_INVESTMENT,
    payload: post
  };
}

export function setInvestments(investments, userId, index) {
  return {
    type: SET_INVESTMENTS,
    payload: {
      investments,
      userId,
      index
    }
  };
}

export function loadingInvestments() {
  return {
    type: LOADING_INVESTMENTS
  };
}

export function setPostInvestments(data, postId, skip) {
  return {
    type: SET_POST_INVESTMENTS,
    payload: {
      postId,
      data,
      index: skip || 0
    }
  };
}

export function setUsers(users) {
  return {
    type: SET_USERS,
    payload: users
  };
}

// optimistic ui
export function vote({ amount, post, user, vote: undo, displayBetPrompt }) {
  return async dispatch => {
    try {
      if (undo) dispatch(undoPostVote(post._id));
      else dispatch(updatePostVote({ post: post._id, amount }));
      const res = await dispatch(
        api.request({
          method: 'POST',
          endpoint: 'invest',
          path: '/',
          body: JSON.stringify({
            investor: user._id,
            amount,
            post
          })
        })
      );
      if (res.undoInvest) dispatch(undoPostVote(post._id));
      else dispatch(updatePostVote(res.investment));
      const isComment = !!post.parentPost;
      if (amount > 0 && !undo) {
        const showingPushBanner = await dispatch(
          showPushNotificationPrompt({
            type: isComment ? 'upvoteComment' : 'upvotePost'
          })
        );
        if (!showingPushBanner && displayBetPrompt) dispatch(showBetPrompt);
      }
      return res;
    } catch (error) {
      if (undo) dispatch(updatePostVote({ post: post._id, amount }));
      else dispatch(undoPostVote(post._id));
      return Alert.alert(error.message);
    }
  };
}

export function getInvestments(userId, skip, limit) {
  return async dispatch => {
    try {
      dispatch(loadingInvestments());
      const res = await dispatch(
        api.request({
          method: 'GET',
          endpoint: 'invest',
          path: '/' + userId,
          query: { skip, limit }
        })
      );
      const data = normalize({ investments: res }, { investments: [investmentSchema] });
      dispatch(setPostsSimple(data));
      dispatch(setInvestments(data, userId, skip));
    } catch (err) {
      // console.log(err);
    }
  };
}

export function loadingPostInvestments(postId) {
  return {
    type: LOADING_POST_INVESTMENTS,
    payload: postId
  };
}

export function getPostInvestments(postId, limit, skip) {
  return async dispatch => {
    try {
      dispatch(loadingPostInvestments(postId));
      const res = await dispatch(
        api.request({
          method: 'GET',
          endpoint: 'invest',
          query: { skip, limit },
          path: `/post/${postId}`
        })
      );
      const data = normalize({ investments: res }, { investments: [investmentSchema] });
      dispatch(setUsers(data.entities.users));
      dispatch(setPostInvestments(data, postId, skip));
    } catch (err) {
      // console.warn(err);
    }
  };
}

export function bet({ postId, stakedTokens }) {
  return async dispatch => {
    try {
      const res = await dispatch(
        api.request({
          method: 'POST',
          endpoint: 'invest',
          path: '/bet',
          body: JSON.stringify({
            postId,
            stakedTokens
          })
        })
      );
      return dispatch(updatePostVote(res));
    } catch (err) {
      Alert.alert(err.message);
      throw err;
    }
  };
}
