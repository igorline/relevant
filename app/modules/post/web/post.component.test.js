import React from 'react';
import renderer from 'react-test-renderer';
import { Post } from 'modules/post/web/post.component';
import 'jest-styled-components';
import { MemoryRouter } from 'react-router-dom';
import { auth, post1, usersState } from 'app/utils/testData';

const props = {
  post: post1,
  link: post1.link,
  actions: {},
  usersState,
  auth,
  location: { pathname: 'post/' + post1._id },
  history: { push: jest.fn() }
};

test('Snapshot Post Web', () => {
  const tree = renderer.create(
    <MemoryRouter>
      <Post {...props} />
    </MemoryRouter>).toJSON();
  expect(tree).toMatchSnapshot();
});
