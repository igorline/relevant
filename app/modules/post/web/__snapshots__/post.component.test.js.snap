// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Post Web 1`] = `
.c0 {
  color: #9C9C9C;
  font-size: 10px;
  font-family: Arial,sans-serif;
}

.c0 a {
  color: #9C9C9C;
}

<div
  className="post"
  onClick={[Function]}
>
  <a
    href="https://example.com/testPost1"
    rel="noopener noreferrer"
    target="_blank"
  >
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="minimalPost postinfo  "
      >
        <div
          className="image"
        >
          <div
            style={
              Object {
                "backgroundImage": "url(link_img1.jpg)",
              }
            }
          />
        </div>
        <div
          className="headlineContainer"
        >
          <h3
            className="headline bebasRegular"
          >
            postTitle1
          </h3>
          <div
            className="domain"
          >
            link.domain1
          </div>
        </div>
      </div>
    </div>
  </a>
  <div
    className="postContent"
  >
    <div
      style={
        Object {
          "flex": 1,
          "minWidth": 0,
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <div
          className="avatarBox "
        >
          <a
            className="avatar"
            href="/user/profile/handle1"
            onClick={[Function]}
            style={
              Object {
                "backgroundImage": "url(userImage1.jpg)",
                "height": undefined,
                "width": undefined,
              }
            }
          >
            Name1
          </a>
          <div
            className="userBox"
          >
            <div
              className="bebasRegular username"
            >
              <a
                href="/user/profile/handle1"
                onClick={[Function]}
              >
                Name1
              </a>
              <span>
                <span
                  alt="R"
                  className="r"
                  style={
                    Object {
                      "backgroundImage": "url(/img/r-emoji.png)",
                    }
                  }
                />
                11
              </span>
            </div>
            <div
              className="c0"
            >
              <a
                href="/user/profile/handle1"
                onClick={[Function]}
              >
                @
                handle1
              </a>
               â€¢ 0sec ago
            </div>
          </div>
        </div>
        <a
          className="popup"
          onClick={[Function]}
        >
          <span
            className="optionDots"
          >
            ...
          </span>
        </a>
      </div>
      <div
        className="postBody"
      >
        <div
          className=""
        >
          <pre>
            awesome post #1!
          </pre>
           
          <a
            href="/testCommunity/top/tag11"
            onClick={[Function]}
          >
            #
            tag11
          </a>
          <a
            href="/testCommunity/top/tag12"
            onClick={[Function]}
          >
            #
            tag12
          </a>
        </div>
        <div
          className="postbuttons"
        >
          <div
            className="left"
          >
            <a
              onClick={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <img
                alt="Upvote"
                className="upvote"
                src="/img/upvote.png"
              />
            </a>
            <div
              className="fraction"
            >
              <div
                className="dem"
              >
                <img
                  alt="R"
                  src="/img/r-gray.svg"
                />
                45
              </div>
            </div>
            <a
              onClick={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <img
                alt="Downvote"
                className="downvote"
                src="/img/downvote-gray.svg"
              />
            </a>
          </div>
          <div
            className="right"
          >
            <a
              className="commentcount details"
              href="/post/1111"
              onClick={[Function]}
            >
              <img
                alt="Comment"
                src="/img/comment.svg"
              />
              <span>
                 comment
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
